aspnetcore/test/middleware.md

```diff
@@ -0,0 +1,104 @@
+
+[!code-csharp[](middleware/samples_snapshot/3.x/false-failure-check.cs?highlight=22)]
+
+Change the assertion to test the middleware under normal operating conditions. The final test uses `Assert.Equal( ... );`. Run the test again to confirm that it passes.
+
+[!code-csharp[](middleware/samples_snapshot/3.x/final-test.cs?highlight=22)]
+
+## Send requests with HttpContext
+
+A test app can also send a request using [SendAsync(Action\<HttpContext>, CancellationToken)](xref:Microsoft.AspNetCore.TestHost.TestServer.SendAsync%2A). In the following example, several checks are made when `https://example.com/A/Path/?and=query` is processed by the middleware:
+
+```csharp
+[Fact]
+public async Task TestMiddleware_ExpectedResponse()
+{
+    using var host = await new HostBuilder()
+        .ConfigureWebHost(webBuilder =>
+        {
+            webBuilder
+                .UseTestServer()
+                .ConfigureServices(services =>
+                {
+                    services.AddMyServices();
+                })
+                .Configure(app =>
+                {
+                    app.UseMiddleware<MyMiddleware>();
+                });
+        })
+        .StartAsync();
+
+    var server = host.GetTestServer();
+    server.BaseAddress = new Uri("https://example.com/A/Path/");
+
+    var context = await server.SendAsync(c =>
+    {
+        c.Request.Method = HttpMethods.Post;
+        c.Request.Path = "/and/file.txt";
+        c.Request.QueryString = new QueryString("?and=query");
+    });
+
+    Assert.True(context.RequestAborted.CanBeCanceled);
+    Assert.Equal(HttpProtocol.Http11, context.Request.Protocol);
+    Assert.Equal("POST", context.Request.Method);
+    Assert.Equal("https", context.Request.Scheme);
+    Assert.Equal("example.com", context.Request.Host.Value);
+    Assert.Equal("/A/Path", context.Request.PathBase.Value);
+    Assert.Equal("/and/file.txt", context.Request.Path.Value);
+    Assert.Equal("?and=query", context.Request.QueryString.Value);
+    Assert.NotNull(context.Request.Body);
+    Assert.NotNull(context.Request.Headers);
+    Assert.NotNull(context.Response.Headers);
+    Assert.NotNull(context.Response.Body);
+    Assert.Equal(404, context.Response.StatusCode);
+    Assert.Null(context.Features.Get<IHttpResponseFeature>().ReasonPhrase);
+}
+```
+
+<xref:Microsoft.AspNetCore.TestHost.TestServer.SendAsync%2A> permits direct configuration of an <xref:Microsoft.AspNetCore.Http.HttpContext> object rather than using the <xref:Microsoft.AspNetCore.Http.HttpContext> abstractions. Use <xref:Microsoft.AspNetCore.TestHost.TestServer.SendAsync%2A> to manipulate structures only available on the server, such as [HttpContext.Items](xref:Microsoft.AspNetCore.Http.HttpContext.Items) or [HttpContext.Features](xref:Microsoft.AspNetCore.Http.HttpContext.Features).
```

Comment:
```suggestion
<xref:Microsoft.AspNetCore.TestHost.TestServer.SendAsync%2A> permits direct configuration of an <xref:Microsoft.AspNetCore.Http.HttpContext> object rather than using the <xref:System.Net.Http.HttpClient> abstractions. Use <xref:Microsoft.AspNetCore.TestHost.TestServer.SendAsync%2A> to manipulate structures only available on the server, such as [HttpContext.Items](xref:Microsoft.AspNetCore.Http.HttpContext.Items) or [HttpContext.Features](xref:Microsoft.AspNetCore.Http.HttpContext.Features).
```